File: web/src/math/spokeLength.ts
export type Side = "left" | "right";

export function lacingAngleRad(crosses: number, totalSpokes: number): number {
  if (totalSpokes <= 0) throw new Error("total_spokes must be positive");
  if (totalSpokes % 2) throw new Error("total_spokes must be even");
  return (2 * Math.PI * crosses) / (totalSpokes / 2);
}

/**
 * Classic crossed-spoke length with rim at the origin in the wheel plane.
 * @param flangeOffsetMm Axial distance (mm) from the **wheel center plane**
 *   (hub mid-width: halfway between outer hub faces) to this flange’s spoke hole
 *   circle — not x/y from dropouts unless converted (see {@link flangeOffsetsFromHubOverallWidth}).
 */
export function spokeLengthMm(
  erdMm: number,
  flangeRadiusMm: number,
  flangeOffsetMm: number,
  crosses: number,
  totalSpokes: number,
): number {
  const R = erdMm / 2;
  const r = flangeRadiusMm;
  const w = flangeOffsetMm;
  const alpha = lacingAngleRad(crosses, totalSpokes);
  let under = R * R + r * r + w * w - 2 * R * r * Math.cos(alpha);
  if (under < 0) under = 0;
  return Math.sqrt(under);
}

export interface SpokeResult {
  index: number;
  side: Side;
  lengthMm: number;
  rimAngleRad: number;
  hubAngleRad: number;
}

export interface FlangeOffsetsFromHubWidth {
  halfWidthMm: number;
  leftFlangeOffsetMm: number;
  rightFlangeOffsetMm: number;
  flangeToFlangeMm: number;
}

export function flangeOffsetsFromHubOverallWidth(
  overallWidthMm: number,
  leftOuterToLeftFlangeMm: number,
  rightOuterToRightFlangeMm: number,
): FlangeOffsetsFromHubWidth {
  if (overallWidthMm <= 0) throw new Error("overall_width_mm must be positive");
  const h = overallWidthMm / 2;
  const x = leftOuterToLeftFlangeMm;
  const y = rightOuterToRightFlangeMm;
  if (x < 0 || y < 0) throw new Error("x and y must be non-negative");
  const L = h - x;
  const R = h - y;
  if (L < 0 || R < 0) {
    throw new Error(
      "x and y cannot exceed half the hub width (negative L or R)",
    );
  }
  return {
    halfWidthMm: h,
    leftFlangeOffsetMm: L,
    rightFlangeOffsetMm: R,
    flangeToFlangeMm: L + R,
  };
}

export function maxCrosses(totalSpokes: number): number {
  if (totalSpokes < 4) return 0;
  return Math.max(0, Math.floor((totalSpokes - 1) / 4));
}

/**
 * Approximate J-bend head clearance at the flange: arc pitch between holes
 * × cos(lacing angle), minus effective hole intrusion (common calculator model).
 * Matches typical wheel-building spreadsheets when hole diameter is hub drilling.
 */
export function spokeHeadClearanceApproxMm(params: {
  flangePcdMm: number;
  spokeCount: number;
  crosses: number;
  flangeHoleDiameterMm: number;
}): number {
  const nHalf = params.spokeCount / 2;
  const pitch = (Math.PI * params.flangePcdMm) / nHalf;
  const alpha = lacingAngleRad(params.crosses, params.spokeCount);
  return pitch * Math.cos(alpha) - params.flangeHoleDiameterMm;
}

/**
 * Rim entry angle: angle in the wheel plane between the spoke and the rim radius
 * at the nipple (first spoke of each side: left index 0, right index 1).
 */
export function rimEntryAngleDeg(params: {
  erdMm: number;
  flangeRadiusMm: number;
  crosses: number;
  spokeCount: number;
  side: Side;
}): number {
  const R = params.erdMm / 2;
  const r = params.flangeRadiusMm;
  const alpha = lacingAngleRad(params.crosses, params.spokeCount);
  const rimIndex = params.side === "left" ? 0 : 1;
  const phi = (2 * Math.PI * rimIndex) / params.spokeCount;
  const hubPhi = params.side === "left" ? phi - alpha : phi + alpha;
  const rmx = R * Math.cos(phi);
  const rmy = R * Math.sin(phi);
  const hx = r * Math.cos(hubPhi) - rmx;
  const hy = r * Math.sin(hubPhi) - rmy;
  const horiz = Math.hypot(hx, hy);
  if (horiz < 1e-9) return 0;
  const dot = hx * -Math.cos(phi) + hy * -Math.sin(phi);
  const c = Math.min(1, Math.max(-1, dot / horiz));
  return (Math.acos(c) * 180) / Math.PI;
}

export function buildSpokeResults(params: {
  erdMm: number;
  spokeCount: number;
  crosses: number;
  leftFlangeRadiusMm: number;
  rightFlangeRadiusMm: number;
  leftFlangeOffsetMm: number;
  rightFlangeOffsetMm: number;
  flangeHoleDiameterMm?: number;
  nippleCorrectionMm?: number;
  rotationRad?: number;
}): SpokeResult[] {
  const {
    erdMm,
    spokeCount: n,
    crosses,
    leftFlangeRadiusMm,
    rightFlangeRadiusMm,
    leftFlangeOffsetMm,
    rightFlangeOffsetMm,
    flangeHoleDiameterMm = 0,
    nippleCorrectionMm = 0,
    rotationRad = 0,
  } = params;
  const alpha = lacingAngleRad(crosses, n);
  const out: SpokeResult[] = [];
  for (let i = 0; i < n; i++) {
    const side: Side = i % 2 === 0 ? "left" : "right";
    const rFl = side === "left" ? leftFlangeRadiusMm : rightFlangeRadiusMm;
    const wFl = side === "left" ? leftFlangeOffsetMm : rightFlangeOffsetMm;
    const base = (2 * Math.PI * i) / n;
    const phi = base + rotationRad;
    const hubPhi = side === "left" ? phi - alpha : phi + alpha;
    const raw = spokeLengthMm(erdMm, rFl, wFl, crosses, n);
    const length =
      raw - flangeHoleDiameterMm / 2 + nippleCorrectionMm;
    out.push({
      index: i,
      side,
      lengthMm: length,
      rimAngleRad: phi,
      hubAngleRad: hubPhi,
    });
  }
  return out;
}

File: legacy/core/spoke_length.py
"""
Spoke length from rim ERD, hub flange radius, axial offset, crosses, and spoke count.

Uses the usual planar cosine law (rim circle / hub hole circle) extended with axial
distance *w* (flange offset), equivalent to either of:

    horizontal² = R² + r² − 2·R·r·cos(α)
    L_raw       = sqrt(horizontal² + w²)

    L_raw = sqrt(R² + r² + w² − 2·R·r·cos(α))   # same value, expanded

with α = 4π·C/N = 2π·C/(N/2) radians (C = crosses, N = total spokes), same as
``theta = (2*math.pi*crosses) / (total_spokes//2)`` in two-step formulas.

After *L_raw*, many calculators subtract half the **hub spoke hole diameter**
(spoke passes through the hole; ordering length is shorter). Nipple / rim
corrections are separate (see form).

References: standard wheel-building geometry (e.g. Damon Rinard / Roger Musson style
treatments of the hub–rim triangle).
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Literal

Side = Literal["left", "right"]


def lacing_angle_rad(crosses: int, total_spokes: int) -> float:
    """α in the hub plane; equals ``2π·crosses / (total_spokes/2)`` (per-side index)."""
    if total_spokes <= 0:
        raise ValueError("total_spokes must be positive")
    if total_spokes % 2:
        raise ValueError("total_spokes must be even")
    return (2.0 * math.pi * crosses) / (total_spokes // 2)


def spoke_length_mm(
    erd_mm: float,
    flange_radius_mm: float,
    flange_offset_mm: float,
    crosses: int,
    total_spokes: int,
) -> float:
    """Straight-line distance (mm) from nipple seat circle (ERD) to flange hole.

    ``flange_offset_mm`` is axial distance from the wheel center plane (rim /
    nipple plane when the rim is centered) to this flange's hole circle — the
    usual left/right flange offset (e.g. *L* and *R* on hub diagrams).
    """
    R = erd_mm / 2.0
    r = flange_radius_mm
    w = flange_offset_mm
    alpha = lacing_angle_rad(crosses, total_spokes)
    under = R * R + r * r + w * w - 2.0 * R * r * math.cos(alpha)
    if under < 0.0:
        under = 0.0
    return math.sqrt(under)


@dataclass(frozen=True)
class SpokeResult:
    index: int
    side: Side
    length_mm: float
    rim_angle_rad: float
    hub_angle_rad: float


@dataclass(frozen=True)
class FlangeOffsetsFromHubWidth:
    """Flange offsets *L* / *R* and flange spacing *F* from overall hub width and *x* / *y*.

    Convention (common on hub drawings): overall width is locknut-to-locknut (or outer
    faces where the hub meets the frame). *x* = from the **left** outer face to the
    **left** flange hole-circle center; *y* = from the **right** outer face to the
    **right** flange center. The wheel center plane is **midway** between those faces
    (half overall width *h*). Then *L* = *h* − *x*, *R* = *h* − *y*, *F* = *L* + *R*.
    """

    half_width_mm: float
    left_flange_offset_mm: float
    right_flange_offset_mm: float
    flange_to_flange_mm: float


def flange_offsets_from_hub_overall_width(
    overall_width_mm: float,
    left_outer_to_left_flange_mm: float,
    right_outer_to_right_flange_mm: float,
) -> FlangeOffsetsFromHubWidth:
    """
    Compute *L*, *R*, and *F* from hub overall width and edge-to-flange dimensions.

    Raises ``ValueError`` if inputs are inconsistent (negative *L* or *R*).
    """
    if overall_width_mm <= 0:
        raise ValueError("overall_width_mm must be positive")
    h = overall_width_mm / 2.0
    x = float(left_outer_to_left_flange_mm)
    y = float(right_outer_to_right_flange_mm)
    if x < 0 or y < 0:
        raise ValueError("x and y must be non-negative")
    L = h - x
    R = h - y
    if L < 0 or R < 0:
        raise ValueError(
            "x and y cannot exceed half the hub width (negative L or R)"
        )
    F = L + R
    return FlangeOffsetsFromHubWidth(
        half_width_mm=h,
        left_flange_offset_mm=L,
        right_flange_offset_mm=R,
        flange_to_flange_mm=F,
    )


def max_crosses(total_spokes: int) -> int:
    """Upper bound for crosses so holes do not overlap tangentially (guard rail)."""
    if total_spokes < 4:
        return 0
    # require 4*C < N so indexing stays in valid tangential range
    return max(0, (total_spokes - 1) // 4)


def build_spoke_results(
    *,
    erd_mm: float,
    spoke_count: int,
    crosses: int,
    left_flange_radius_mm: float,
    right_flange_radius_mm: float,
    left_flange_offset_mm: float,
    right_flange_offset_mm: float,
    flange_hole_diameter_mm: float = 0.0,
    nipple_correction_mm: float = 0.0,
    rotation_rad: float = 0.0,
) -> list[SpokeResult]:
    """
    One result per spoke. Index 0..N-1 around the rim; even index = left, odd = right.
    rotation_rad spins the whole pattern (e.g. align a valve hole) CCW from +x.
    hub_angle_rad: hub hole toward which the spoke runs (top-view plan).
    """
    n = spoke_count
    alpha = lacing_angle_rad(crosses, n)
    out: list[SpokeResult] = []
    for i in range(n):
        side: Side = "left" if i % 2 == 0 else "right"
        r_fl = left_flange_radius_mm if side == "left" else right_flange_radius_mm
        w_fl = left_flange_offset_mm if side == "left" else right_flange_offset_mm
        base = 2.0 * math.pi * i / n
        phi = base + rotation_rad
        # Left flange holes “lag” vs right in a common convention; keeps diagram readable.
        if side == "left":
            hub_phi = phi - alpha
        else:
            hub_phi = phi + alpha
        raw = spoke_length_mm(erd_mm, r_fl, w_fl, crosses, n)
        length = raw - (flange_hole_diameter_mm / 2.0) + nipple_correction_mm
        out.append(
            SpokeResult(
                index=i,
                side=side,
                length_mm=length,
                rim_angle_rad=phi,
                hub_angle_rad=hub_phi,
            )
        )
    return out
